File: data_structures/go/hashMaps.go
package main

import (
	"fmt"
	"hash/fnv"
)

type KeyValue struct {
	key   interface{}
	value interface{}
}

type HashMap struct {
	buckets  [][]KeyValue
	capacity int
}

func NewHashMap(capacity int) *HashMap {
	return &HashMap{
		buckets:  make([][]KeyValue, max(1, capacity)),
		capacity: max(1, capacity),
	}
}

// Insertion
func (h *HashMap) SetValue(value interface{}, key interface{}) {
	index := h.bucketIndex(key)
	for i, kv := range h.buckets[index] {
		if kv.key == key {
			h.buckets[index][i].value = value
			return
		}
	}
	h.buckets[index] = append(h.buckets[index], KeyValue{key, value})
}

// Retrieval
func (h *HashMap) GetValue(key interface{}) interface{} {
	index := h.bucketIndex(key)
	for _, kv := range h.buckets[index] {
		if kv.key == key {
			return kv.value
		}
	}
	return nil
}

// Removal
func (h *HashMap) RemoveValue(key interface{}) {
	index := h.bucketIndex(key)
	var updated []KeyValue
	for _, kv := range h.buckets[index] {
		if kv.key != key {
			updated = append(updated, kv)
		}
	}
	h.buckets[index] = updated
}

// Count
func (h *HashMap) Count() int {
	totalCount := 0
	for _, bucket := range h.buckets {
		totalCount += len(bucket)
	}
	return totalCount
}

// Keys and Values
func (h *HashMap) AllKeys() []interface{} {
	var keys []interface{}
	for _, bucket := range h.buckets {
		for _, kv := range bucket {
			keys = append(keys, kv.key)
		}
	}
	return keys
}

func (h *HashMap) AllValues() []interface{} {
	var values []interface{}
	for _, bucket := range h.buckets {
		for _, kv := range bucket {
			values = append(values, kv.value)
		}
	}
	return values
}

// Helper Functions
func (h *HashMap) IsEmpty() bool {
	for _, bucket := range h.buckets {
		if len(bucket) > 0 {
			return false
		}
	}
	return true
}

func (h *HashMap) RemoveAll() {
	h.buckets = make([][]KeyValue, h.capacity)
}

func (h *HashMap) Contains(key interface{}) bool {
	index := h.bucketIndex(key)
	for _, kv := range h.buckets[index] {
		if kv.key == key {
			return true
		}
	}
	return false
}

func (h *HashMap) LoadFactor() float64 {
	return float64(h.Count()) / float64(h.capacity)
}

func (h *HashMap) KeyValuePairs() map[interface{}]interface{} {
	pairs := make(map[interface{}]interface{})
	for _, bucket := range h.buckets {
		for _, kv := range bucket {
			pairs[kv.key] = kv.value
		}
	}
	return pairs
}

// Update Values
func (h *HashMap) UpdateValue(value interface{}, key interface{}) {
	index := h.bucketIndex(key)
	for i, kv := range h.buckets[index] {
		if kv.key == key {
			h.buckets[index][i].value = value
			return
		}
	}
	h.buckets[index] = append(h.buckets[index], KeyValue{key, value})
}

func (h *HashMap) UpdateValues(dictionary map[interface{}]interface{}) {
	for key, value := range dictionary {
		h.UpdateValue(value, key)
	}
}

// ... Add more hashmap operations as needed ...

func (h *HashMap) bucketIndex(key interface{}) int {
	hash := fnv.New32a()
	fmt.Fprintf(hash, "%v", key)
	return int(hash.Sum32()) % h.capacity
}

File: data_structures/swift/hashMaps.swift
struct KeyValue<T: Hashable, U> {
    let key: T
    var value: U
}

struct HashMap<T: Hashable, U> {
    private var buckets: [[KeyValue<T, U>]]
    private var capacity: Int

    init(capacity: Int) {
        self.capacity = max(1, capacity)
        self.buckets = Array(repeating: [], count: self.capacity)
    }

    // MARK: - Insertion

    /// Insert a value for a given key.
    mutating func setValue(_ value: U, forKey key: T) {
        let index = bucketIndex(for: key)
        for (i, kv) in buckets[index].enumerated() {
            if kv.key == key {
                buckets[index][i].value = value
                return
            }
        }
        buckets[index].append(KeyValue(key: key, value: value))
    }

    // MARK: - Retrieval

    /// Get the value for a given key.
    func getValue(forKey key: T) -> U? {
        let index = bucketIndex(for: key)
        for kv in buckets[index] {
            if kv.key == key {
                return kv.value
            }
        }
        return nil
    }

    // MARK: - Removal

    /// Remove the value for a given key.
    mutating func removeValue(forKey key: T) {
        let index = bucketIndex(for: key)
        buckets[index] = buckets[index].filter { $0.key != key }
    }

    // MARK: - Count

    /// Get the total count of elements in the hashmap.
    func count() -> Int {
        return buckets.reduce(0) { $0 + $1.count }
    }

    // MARK: - Keys and Values

    /// Get an array of all keys in the hashmap.
    func allKeys() -> [T] {
        return buckets.flatMap { $0.map { $0.key } }
    }

    /// Get an array of all values in the hashmap.
    func allValues() -> [U] {
        return buckets.flatMap { $0.map { $0.value } }
    }

    // MARK: - Helper Functions

    /// Check if the hashmap is empty.
    func isEmpty() -> Bool {
        return buckets.allSatisfy { $0.isEmpty }
    }

    /// Clear all key-value pairs in the hashmap.
    mutating func removeAll() {
        buckets = Array(repeating: [], count: capacity)
    }

    /// Check if a key exists in the hashmap.
    func contains(key: T) -> Bool {
        return buckets[bucketIndex(for: key)].contains { $0.key == key }
    }

    /// Get the load factor of the hashmap.
    func loadFactor() -> Double {
        return Double(count()) / Double(capacity)
    }

    /// Get a set of key-value pairs in the hashmap.
    func keyValuePairs() -> Set<KeyValue<T, U>> {
        var set = Set<KeyValue<T, U>>()
        for bucket in buckets {
            set.formUnion(bucket)
        }
        return set
    }

    // MARK: - Update Values

    /// Update a value for a given key. If the key doesn't exist, it inserts a new key-value pair.
    mutating func updateValue(_ value: U, forKey key: T) {
        let index = bucketIndex(for: key)
        for (i, kv) in buckets[index].enumerated() {
            if kv.key == key {
                buckets[index][i].value = value
                return
            }
        }
        buckets[index].append(KeyValue(key: key, value: value))
    }

    /// Update values for multiple keys using a dictionary.
    mutating func updateValues(_ dictionary: [T: U]) {
        for (key, value) in dictionary {
            updateValue(value, forKey: key)
        }
    }

    // ... Add more hashmap operations as needed ...
}
